proof/refine/ARM/Ipc_R.thy

```diff
@@ -5881,6 +5838,318 @@ lemma si_blk_makes_runnable':
 crunches possibleSwitchTo, completeSignal
   for pred_tcb_at'[wp]: "pred_tcb_at' proj P t"
 
+crunches replyRemove
+  for ksSchedulerAction[wp]: "\<lambda>s. P (ksSchedulerAction s)"
+  (simp: crunch_simps)
+
+crunch invs'[wp]: getSanitiseRegisterInfo invs'
+
+crunches getSanitiseRegisterInfo
+  for it'[wp]: "\<lambda>s. P (ksIdleThread s)"
+  and tcb_at'[wp]: "tcb_at' a"
+
+lemma handleFaultReply_invs[wp]:
+  "\<lbrace>invs' and tcb_at' t\<rbrace> handleFaultReply x t label msg \<lbrace>\<lambda>rv. invs'\<rbrace>"
+  unfolding handleFaultReply_def
+  by (cases x; wpsimp simp: handleArchFaultReply_def split: arch_fault.split)
+
+lemma doReplyTransfer_corres:
+  "corres dc
+     (einvs and reply_at reply and tcb_at sender)
+     invs'
+     (do_reply_transfer sender reply grant)
+     (doReplyTransfer sender reply grant)"
+  supply if_split [split del]
+  apply (simp add: do_reply_transfer_def doReplyTransfer_def cong: option.case_cong)
+  apply (rule stronger_corres_guard_imp)
+    apply (rule corres_split [OF getReply_TCB_corres])
+      apply (simp add: maybeM_def)
+      apply (rule corres_option_split [OF refl corres_return_trivial])
+      apply (rename_tac recv_opt receiverOpt recvr)
+      apply (rule_tac Q="\<lambda>s. einvs s \<and> tcb_at sender s \<and> tcb_at recvr s \<and>
+                             reply_tcb_reply_at (\<lambda>xa. xa = Some recvr) reply s" and
+                      Q'="invs'"
+             in corres_split [OF _ _ gts_sp gts_sp' ])
+       apply (rule gts_corres)
+      apply (case_tac rv; simp add: isReply_def)
```

Comment:
It's hard to tell just from the proof script, but my guess is that it might help to have a rename around here.